### PR TITLE
Added w3d parsing support for Earth & Beyond and Renegade

### PR DESCRIPTION
**Pull request by jpf200124**
Added Support for Earth and Beyond and Renegade W3D's. Most of the ENB parsing was based on my work from the enb-emulator. 

There are still a bunch of unknown's that I'll continue to try and figure out.

While these games did not specifically use the sage engine I think it's important to have the w3d format documented as best as possible and this project seems to be the furthest along with the format.

I re-ran the parsing through generals and zero hour just to make sure there were no parsing errors.

### CODE REVIEW

File: src/OpenSage.Game/Data/W3d/W3dAggregateInfo.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+                return result;
+            });
+        }
+
+        protected override void WriteToOverride(BinaryWriter writer)
+        {
+            writer.WriteFixedLengthString(BaseModelName, W3dConstants.NameLength * 2);
+            writer.Write(SubObjectCount);
+            SubObjects.ForEach(delegate (W3dAggregateSubObject subObject)
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dAggregateInfo.cs]: Please could you change this to a normal `foreach`? We tend to use them for simplicity, and also for performance / memory allocation reasons.

File: src/OpenSage.Game/Data/W3d/W3dAnimation.cs [hunk 2]
```diff
@@ -17,6 +17,9 @@ internal static W3dAnimation Parse(BinaryReader reader, W3dParseContext context)
             {
                 var result = new W3dAnimation();
 
+                // TODO: generals zero hour assets: UISabotr throw chunktype errors.
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dAnimation.cs]: You'll notice `UISabotr_*` files are ignored in `W3dFileTests` for that exact reason. As far as I can tell, they contain corrupt chunks. So I think no need for this comment.

File: src/OpenSage.Game/Data/W3d/W3dChunk.cs [hunk 3]
```diff
@@ -23,7 +23,7 @@ public abstract class W3dChunk
             var result = parseCallback(chunkHeader);
 
             result.StartPosition = startPosition;
-            result.EndPosition = endPosition;
+            result.EndPosition =  endPosition;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dChunk.cs]: Please remove this whitespace change.

File: src/OpenSage.Game/Data/W3d/W3dChunkType.cs [hunk 4]
```diff
@@ -136,6 +136,7 @@ public enum W3dChunkType
             W3D_CHUNK_AGGREGATE_INFO,                                       // references to 'contained' models
             W3D_CHUNK_TEXTURE_REPLACER_INFO,                                    // information about which meshes need textures replaced
             W3D_CHUNK_AGGREGATE_CLASS_INFO,                                 // information about the original class that created this aggregate
+            W3D_CHUNK_AGGREGATE_UNKNOWN = 393728,                           // Unknown chunk found in enb asset ("npc14b.w3d")
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dChunkType.cs]: This seems more like a corrupt chunk. Is the model that uses this ever referenced by the game itself?

File: src/OpenSage.Game/Data/W3d/W3dCollectionPlaceholder.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            return ParseChunk(reader, context, header =>
+            {
+                var result = new W3dCollectionPlaceholder
+                {
+                    TransformA = reader.ReadVector3(),
+                    TransformB = reader.ReadVector3(),
+                    TransformC = reader.ReadVector3(),
+                    TransformD = reader.ReadVector3(),
+                    Name = reader.ReadFixedLengthString(W3dConstants.NameLength),
+                    UnknownBytes = reader.ReadBytes((int)context.CurrentEndPosition - (int)reader.BaseStream.Position)
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dCollectionPlaceholder.cs]: If you do more PRs, please could you add a space after casts?  i.e.

``` csharp
UnknownBytes = reader.ReadBytes((int) context.CurrentEndPosition - (int) reader.BaseStream.Position)
```

Don't worry for this PR though.

File: src/OpenSage.Game/Data/W3d/W3dDazzleName.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                    Name = reader.ReadFixedLengthString((int)context.CurrentEndPosition - (int)reader.BaseStream.Position)
+                };
+
+                return result;
+            });
+        }
+
+        protected override void WriteToOverride(BinaryWriter writer)
+        {
+            writer.Write(Name);
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dDazzleName.cs]: Please use `WriteFixedLengthString`.

File: src/OpenSage.Game/Data/W3d/W3dDazzleTypeName.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                    TypeName = reader.ReadFixedLengthString((int)context.CurrentEndPosition - (int)reader.BaseStream.Position)
+                };
+
+                return result;
+            });
+        }
+
+        protected override void WriteToOverride(BinaryWriter writer)
+        {
+            writer.Write(TypeName);
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dDazzleTypeName.cs]: Please use `WriteFixedLengthString`.

File: src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs [hunk 8]
```diff
@@ -21,7 +21,7 @@ internal static W3dEmitterLineProperties Parse(BinaryReader reader, W3dParseCont
             {
                 var result = new W3dEmitterLineProperties
                 {
-                    Flags = reader.ReadUInt32AsEnum<W3dEmitterLineFlags>(),
+                    Flags = (W3dEmitterLineFlags)reader.ReadUInt32(),
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs]: Why this change? Which new values have you found?

File: src/OpenSage.Game/Data/W3d/W3dHModelHeader.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                // TODO: Determine W3dHModelHeader UnknownBytes
+
+                return result;
+            });
+        }
+
+        protected override void WriteToOverride(BinaryWriter writer)
+        {
+            writer.Write(Version);
+            writer.Write(Name);
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dHModelHeader.cs]: Please use `WriteFixedLengthString`.

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiPass.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+                return result;
+            });
+
+            return parsedChunk;
+        }
+
+        protected override IEnumerable<W3dChunk> GetSubChunksOverride()
+        {
+            yield return MaterialInfo;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiPass.cs]: Is this missing some stuff?

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+                return result;
+            });
+
+            return parsedChunk;
+        }
+
+        protected override IEnumerable<W3dChunk> GetSubChunksOverride()
+        {
+            yield return MaterialInfo;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs]: Is this missing some stuff?

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Numerics;
+using OpenSage.Data.Utilities.Extensions;
+
+namespace OpenSage.Data.W3d
+{
+    public sealed class W3dPrelitLightMapMultiTexture : W3dContainerChunk
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs]: Is it worth making a base class for this and `W3dPrelitLightMapMultiPass` and `W3dPrelitUnlit` and `W3dPrelitVertex`, and moving a lot of stuff into that? Seems like they're basically the same thing?

File: src/OpenSage.Game/Data/W3d/W3dRgbF.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
+﻿using System.IO;
+using System.Runtime.InteropServices;
+
+namespace OpenSage.Data.W3d
+{
+    [StructLayout(LayoutKind.Sequential)]
+    public struct W3dRgbF
+    {
+        public float R { get; private set; }
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dRgbF.cs]: We're actually in the process of removing W3D-specific colour types, so please could you use `ColorRgbF` instead?

File: src/OpenSage.Game/Data/W3d/W3dTextureInfo.cs [hunk 14]
```diff
@@ -33,7 +33,7 @@ internal static W3dTextureInfo Parse(BinaryReader reader, W3dParseContext contex
             {
                 return new W3dTextureInfo
                 {
-                    Attributes = reader.ReadUInt16AsEnumFlags<W3dTextureFlags>(),
+                    Attributes = (W3dTextureFlags) reader.ReadUInt16(),
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dTextureInfo.cs]: Why this change? I'd prefer we add the new values to `W3dTextureFlags`.

File: src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs [hunk 15]
```diff
@@ -57,6 +57,16 @@ public W3dVertexMapperArgs(W3dChunkType chunkType)
         public float UStep;
         public float VStep;
         public float StepsPerSecond;
+        public float Offset;
+        public float Axis;
+        public float UOffset;
+        public float VOffset;
+        public float ClampFix;
+        public float UseReflect;
+        public float VPreSec;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs]: Should this be `VPerSec`?

File: src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs [hunk 16]
```diff
@@ -57,6 +57,16 @@ public W3dVertexMapperArgs(W3dChunkType chunkType)
...
+        public float Period;
+        public float UperSec;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs]: Please name this `UPerSec`.

File: src/OpenSage.Game/Data/W3d/W3dVertexMappingType.cs [hunk 17]
```diff
@@ -60,6 +60,77 @@ public enum W3dVertexMappingType : uint
...
         /// so that the matrix gets set up with the identity settings.
         /// </summary>
-        BumpEnv
+        BumpEnv,
+
+        Mask,
+        Stage1_Uv,
     }
+
+    /*
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dVertexMappingType.cs]: I don't think we need any of this comment? We already have these constants defined elsewhere.

File: src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs (hunk 8 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs]: Looks like I changed this because it was failing to parse based on a new value. I wasn't sure what the flag was for. I'll revert it and add as a UnknownENB value in the enum. Its value is: ... (0x00000002 << 24) + 1. It is possible this was corruption. I have only found this in a few enb assets.

File: src/OpenSage.Game/Data/W3d/W3dTextureInfo.cs (hunk 14 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dTextureInfo.cs]: Looks like I also changed this because it was failing to parse based on a new value.  I'll revert it and add as a UnknownENB value in the enum. Its value is: ... 0x0100;

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs (hunk 12 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs]: Yes, I had meant to do but somehow got sidetracked! I'll add it in the update.

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiPass.cs (hunk 10 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiPass.cs]: Yep ... Thanks for catching this. I'll add in update.

File: src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs (hunk 15 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dVertexMapperArgs.cs]: Hah, This is not a mistake ... enb assets contain a couple incorrect spellings in the mapper args. Case seems to be different as well. I was quite taken back when I saw this.

File: src/OpenSage.Game/Data/W3d/W3dChunkType.cs (hunk 4 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dChunkType.cs]: Seems like this is indeed corrupt chunks. There are only 2 enb assets that contain this corruption and I can find no reference to them being used in the game or the emulator.

File: src/OpenSage.Game/Data/W3d/W3dCollectionPlaceholder.cs (hunk 5 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dCollectionPlaceholder.cs]: I went through and updated everything.

File: src/OpenSage.Game/Data/W3d/W3dDazzleTypeName.cs (hunk 7 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dDazzleTypeName.cs]: Went through and updated all strings.

File: src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs (hunk 12 above)
Comment (jpf200124) [reply src/OpenSage.Game/Data/W3d/W3dPrelitLightMapMultiTexture.cs]: See PrelitBase

**Comment by jpf200124** [comment]
> Just so you know, even before this PR, I planned to extract the w3d parser to its own project, still within the OpenSAGE repo, probably named OpenSage.FileFormats.W3d. I'll wait until after this PR is merged to do that, so you don't need to redo all this work.

This is Great! I was going to ask if this was something you are planning. I am planning on making a project to do w3d to glTF and was hoping to have this in a separate project from the entire sage engine

File: src/OpenSage.Game/Data/W3d/W3dCollectionHeader.cs [hunk 18]
```diff
@@ -33,7 +35,7 @@ internal static W3dCollectionHeader Parse(BinaryReader reader, W3dParseContext c
         protected override void WriteToOverride(BinaryWriter writer)
         {
             writer.Write(Version);
-            writer.Write(Name);
+            writer.WriteFixedLengthString(Name, NameSize);
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dCollectionHeader.cs]: Sorry - I should have been clearer. No need to store `NameSize`. You can do

``` csharp
writer.WriteFixedLengthString(Name, Name.Length + 1);
```

This way of doing it has come up a few times in w3d chunks now, so we should probably have a helper method for it.

File: src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs [hunk 19]
```diff
@@ -21,7 +21,7 @@ internal static W3dEmitterLineProperties Parse(BinaryReader reader, W3dParseCont
             {
                 var result = new W3dEmitterLineProperties
                 {
-                    Flags = (W3dEmitterLineFlags)reader.ReadUInt32(),
+                    Flags = reader.ReadUInt32AsEnum<W3dEmitterLineFlags>(),
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dEmitterLineProperties.cs]: On further reflection, this should be:

``` csharp
Flags = reader.ReadUInt32AsEnumFlags<W3dEmitterLineFlags>()
```

and then you can remove the new value from `W3dEmitterLineFlags`.

File: src/OpenSage.Game/Data/W3d/W3dPrelitBase.cs [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+using System.IO;
+using System.Numerics;
+using OpenSage.Data.Utilities.Extensions;
+
+namespace OpenSage.Data.W3d
+{
+    public abstract class W3dPrelitBase<TDerived> : W3dContainerChunk
+        where TDerived : W3dPrelitBase<TDerived>, new()
+    {
+        public override W3dChunkType ChunkType { get; } = W3dChunkType.W3D_CHUNK_PRELIT_VERTEX;
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dPrelitBase.cs]: Please remove this override, since it's overridden in all the subclasses.

File: src/OpenSage.Game/Data/W3d/W3dRgbF.cs [hunk 21]
```diff
@@ -1,32 +1,27 @@
-﻿using System.IO;
+﻿using OpenSage.Mathematics;
+using System.IO;
 using System.Runtime.InteropServices;
 
 namespace OpenSage.Data.W3d
 {
-    [StructLayout(LayoutKind.Sequential)]
-    public struct W3dRgbF
+    public static class W3dRgbF
```
Comment (tgjones) [inline src/OpenSage.Game/Data/W3d/W3dRgbF.cs]: There are already extension methods for `ReadColorRgbF` and `Write(ColorRgbF)`, so actually you can remove this whole class.

**Comment by tgjones** [comment]
Thank you, great work!
